File: src/test/java/io/lsdconsulting/lsd/distributed/interceptor/captor/http/derive/SourceTargetDeriverShould.java
package io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.apache.commons.lang3.RandomStringUtils.randomAlphabetic;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;

class SourceTargetDeriverShould {

    private final PropertyServiceNameDeriver propertyServiceNameDeriver = mock(PropertyServiceNameDeriver.class);

    private final SourceTargetDeriver underTest = new SourceTargetDeriver(propertyServiceNameDeriver);

    private final String appName = randomAlphabetic(20);
    private final String sourceName = randomAlphabetic(20);
    private final String targetName = randomAlphabetic(20);
    private final String path = randomAlphabetic(20);

    @BeforeEach
    public void setup() {
        given(propertyServiceNameDeriver.getServiceName()).willReturn(appName);
        underTest.initialise();
    }

    @Test
    void useHeaderForSourceName() {
        final String serviceName = underTest.deriveServiceName(Map.of("Source-Name", List.of(sourceName)));

        assertThat(serviceName, is(sourceName));
    }

    @Test
    void fallbackToPropertyServiceNameIfSourceHeaderMissingValue() {
        final String serviceName = underTest.deriveServiceName(Map.of("Source-Name", List.of()));

        assertThat(serviceName, is(appName));
    }

    @Test
    void fallbackToPropertyServiceNameWhenNoSourceHeader() {
        final String serviceName = underTest.deriveServiceName(Map.of());

        assertThat(serviceName, is(appName));
    }

    @Test
    void useHeaderForTargetName() {
        final String serviceName = underTest.deriveTarget(Map.of("Target-Name", List.of(targetName)), path);

        assertThat(serviceName, is(targetName));
    }

    @Test
    void fallbackToPathTargetHeaderMissingValue() {
        final String serviceName = underTest.deriveTarget(Map.of("Target-Name", List.of()), path);

        assertThat(serviceName, is(path));
    }

    @Test
    void fallbackToPathTargetIfTargetHeaderMissingValueAndNoPath() {
        final String serviceName = underTest.deriveTarget(Map.of("Target-Name", List.of()), null);

        assertThat(serviceName, is("UNKNOWN_TARGET"));
    }

    @Test
    void fallbackToUnknownWhenNoTargetHeader() {
        final String serviceName = underTest.deriveTarget(Map.of(), null);

        assertThat(serviceName, is("UNKNOWN_TARGET"));
    }
}

File: src/test/kotlin/io/lsdconsulting/lsd/distributed/interceptor/captor/http/derive/SourceTargetDeriverShould.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive

import io.lsdconsulting.lsd.distributed.interceptor.captor.common.PropertyServiceNameDeriver
import io.mockk.every
import io.mockk.mockk
import org.apache.commons.lang3.RandomStringUtils.randomAlphabetic
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.`is`
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class SourceTargetDeriverShould {
    private val propertyServiceNameDeriver = mockk<PropertyServiceNameDeriver>()

    private val underTest = SourceTargetDeriver(propertyServiceNameDeriver)

    private val appName = randomAlphabetic(20)
    private val sourceName = randomAlphabetic(20)
    private val targetName = randomAlphabetic(20)
    private val path = randomAlphabetic(20)

    @BeforeEach
    fun setup() {
        every { propertyServiceNameDeriver.serviceName }  returns appName
        underTest.initialise()
    }

    @Test
    fun useHeaderForSourceName() {
        val result = underTest.deriveServiceName(mapOf("Source-Name" to listOf(sourceName)))
        assertThat(result, `is`(sourceName))
    }

    @Test
    fun fallbackToPropertyServiceNameIfSourceHeaderMissingValue() {
        val result = underTest.deriveServiceName(mapOf<String, List<String?>>("Source-Name" to listOf<String>()))
        assertThat(result, `is`(appName))
    }

    @Test
    fun fallbackToPropertyServiceNameWhenNoSourceHeader() {
        val result = underTest.deriveServiceName(mapOf())
        assertThat(result, `is`(appName))
    }

    @Test
    fun useHeaderForTargetName() {
        val result = underTest.deriveTarget(mapOf("Target-Name" to listOf(targetName)), path)
        assertThat(result, `is`(targetName))
    }

    @Test
    fun fallbackToPathTargetHeaderMissingValue() {
        val result = underTest.deriveTarget(mapOf<String, List<String?>>("Target-Name" to listOf<String>()), path)
        assertThat(result, `is`(path))
    }

    @Test
    fun fallbackToPathTargetIfTargetHeaderMissingValueAndNoPath() {
        val result = underTest.deriveTarget(mapOf<String, List<String?>>("Target-Name" to listOf<String>()), null)
        assertThat(result, `is`("UNKNOWN_TARGET"))
    }

    @Test
    fun fallbackToUnknownWhenNoTargetHeader() {
        val result = underTest.deriveTarget(mapOf(), null)
        assertThat(result, `is`("UNKNOWN_TARGET"))
    }
}
